Migrate and fix Optimal control tutorials from qutip-notebooks to qutip-tutorials

rochisha0 opened the pull request:
Migrate and update the Optimal Control tutorials from the `qutip-notebooks` repository to the `qutip-tutorials` repository. Ensure all tutorials are compatible with `qutip-qtrl` instead of the older `qutip.control` module. This involves revising code, validating outputs, and updating instructions to reflect changes functionality. It also removes the `logging` to make the tutorials more compact.

To do:

1. Change the notebooks title

ajgpitch commented:
I checked the `tutorials-v5/optimal-control/control-pulseoptim-Hadamard.ipynb` and it imports `qutip.control`. This is not going work in v5. Needs to use `qutip_qtrl`.

rochisha0 commented:
@ajgpitch I have fixed this and added overview. I think we should figure out, which order we want these notebooks to appear. I can name them in that particular order to show in the website index.

ajgpitch commented:
Definitely looking better now. Still focusing on the Hadamard example...
1. I thought at some point we planned to move (copy) the Hadamard gate function into the main package. It would be nice not to have to depend on qutip_qip - can we check if this is done?
2. It seems that this does not work with the current pip version of qutip-qtrl, do I need to install qutip_qtrl from Github?

rochisha0 commented:
I have removed the dependence on `qutip_qip`. It works for me using the current pip version of `qutip_qtrl`
Software	Version
QuTiP	5.1.0.dev0+0b4260e
Numpy	1.26.4
SciPy	1.13.0
matplotlib	3.9.0
Number of CPUs	8
BLAS Info	Generic
IPython	8.25.0
Python	3.12.3 | packaged by Anaconda, Inc. | (main, May 6 2024, 19:46:43) [GCC 11.2.0]
OS	posix [linux]
Cython	3.0.10
Wed Jan 01 22:48:30 2025 IST

These are the versions I am using.

rochisha0 commented:
@ajgpitch I am not able to understand this error

`E402 module level import not at top of file`

Rest I think everything is fixed

ajgpitch commented:
Maybe it is the empty lines in cell 1 and also perhaps `%matplotlib inline` can be moved after the imports.

rochisha0 commented:
@ajgpitch I have updated the names. I’m considering making additional content changes and can open another PR to integrate this idea properly.

ajgpitch commented:
Tests failing because links need updating in `optimal-control-overview.md`

ajgpitch commented:
Okay, I assume that is because they don't exist yet.
@rochisha0 I will merge this, then you can test.

hodgestar commented:
The qutip-tutorials builds are all broken post this merge.

ajgpitch commented:
Sorry folks. 
@rochisha0 it seems our conjecture re the links was false. Can you please take a look at this urgently and make another PR to fix?
